### PR TITLE
Add support for SMBus and EVGA GTX 1080 FTW

### PR DESCRIPTION
**Pull request by jonasmalacofilho**
Closes: #224

### CODE REVIEW

File: docs/nvidia-guide.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# NVIDIA graphics cards
+_Driver API and source code available in [`liquidctl.driver.nvidia`](../liquidctl/driver/nvidia.py)._
+
+Support for these cards in only available on Linux.
+
+Additional requirements must also be met:
+
+- optional Python dependency `smbus` is available
+- `i2c-dev` kernel module has been loaded
```
Comment (jonasmalacofilho) [inline docs/nvidia-guide.md:9]: I should prepare and include a "load the `i2c-dev` module during boot" configuration file for inclusion in distro packages.

File: docs/nvidia-guide.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+- specific unsafe features have been opted in
+- r/w permissions to card-specific `/dev/i2c-*` devices
```
Comment (jonasmalacofilho) [inline docs/nvidia-guide.md:11]: I also forgot to add a new rule for the EVGA card in [`extra/linux/71-liquidctl.rules`](https://github.com/jonasmalacofilho/liquidctl/blob/master/extra/linux/71-liquidctl.rules).

**Comment by jonasmalacofilho** [comment]
@MarshallAsch 

I know there are a few changes that shouldn't really be part of this PR (despite being inspired by it), sorry about that!  Feel free to ignore those files, but please let me know by adding a "Ignored." comment to the top of each ignored file.

File: docs/developer/style-guide.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+Prefer old-style %-formatting, since this is evaluated lazyly by `logging`, and
+the message will only be formated if the logging level is enabled.
+
+```py
+_LOGGER.warning('value %d, expected %d', current, expected)
+```
+
+When writing informational or debug messages, pay attention to the cost of
+computing each value.  A classic example is hex formatting some bytes, which
+can be expensive; this case can be solved by using
```
Comment (MarshallAsch) [inline docs/developer/style-guide.md:112]: Interesting, I didn't think about that but it makes sense.

File: docs/nvidia-guide.md (hunk 2 above)
Comment (MarshallAsch) [reply docs/nvidia-guide.md:11]: That device is also included PR #228 (I know it should have probably been here but, not a big deal)

File: docs/nvidia-guide.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+---
+
+Jump to a specific card:
+
+* _Series 10/Pascal:_
+    - [EVGA GTX 1080 FTW](#evga-gtx-1080-ftw)
+<!-- - [EVGA GTX 1070 FTW](#evga-gtx-1070-1080-ftw) -->
+<!-- - [EVGA GTX 1080 FTW](#evga-gtx-1070-1080-ftw) -->
```
Comment (MarshallAsch) [inline docs/nvidia-guide.md:20]: Is this line a copy paste error or are those commented out lines different?

File: liquidctl/cli.py [hunk 5]
```diff
@@ -33,6 +33,7 @@
   --single-12v-ocp            Enable single rail +12V OCP
   --pump-mode <mode>          Set the pump mode (certain Corsair coolers)
   --legacy-690lc              Use Asetek 690LC in legacy mode (old Krakens)
+  --non-volatile              Store on non-volatile controller memory
```
Comment (MarshallAsch) [inline liquidctl/cli.py:36]: We will need to remember to add this to the bash completion script in #205

File: liquidctl/driver/nvidia.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        RAINBOW = (0b010, 0)
+        BREATHING = (0b101, 1)
+
+        def __str__(self):
+            return self.name.capitalize()
+
+    @classmethod
+    def probe(cls, smbus, vendor=None, product=None, address=None, match=None,
+              release=None, serial=None, **kwargs):
+        GTX_1080_FTW = 0x6286
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:53]: How come this value is here and not at the top with the other constants?

File: liquidctl/driver/nvidia.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        def __new__(cls, value, required_colors):
+            obj = bytes.__new__(cls, [value])
+            obj._value_ = value
+            obj.required_colors = required_colors
+            return obj
+
+        OFF = (0b000, 0)
+        FIXED = (0b001, 1)
+        RAINBOW = (0b010, 0)
+        BREATHING = (0b101, 1)
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:45]: Are binary values preferred for these constants to hex values?

File: liquidctl/driver/nvidia.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
+"""liquidctl drivers for NVIDIA graphics cards.
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:1]: You sort of mentioned this in one of your messages but why is this driver called `nvidia`? Its intention is not to store the drivers for all NVIDIA graphics cards in this file is it?

File: liquidctl/driver/nvidia.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+        It is possible to store them in non-volatile controller memory by
+        passing `non_volatile=True`.  But as this memory has some unknown yet
+        limited maximum number of write cycles, volatile settings are
+        preferable, if the use case allows for them.
+        """
+
+        check_unsafe('smbus', 'evga_pascal', error=True, **kwargs)
+
+        channel = 'led'
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:132]: should the channel name that is given be checked against the value 'led'? Right now channel can be set to anything and is ignored.

File: liquidctl/driver/smbus.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+        def open(self):
+            """Open the I²C bus."""
+            if not self._smbus:
+                try:
+                    self._smbus = SMBus(self._number)
+                except FileNotFoundError:
+                    if Path('/sys/class/i2c-dev').exists():
+                        raise
+                    raise OSError('kernel module i2c-dev not loaded') from None
```
Comment (MarshallAsch) [inline liquidctl/driver/smbus.py:102]: why is this error `from None` I'm not all too familiar with that syntax. From what I Googled doesn't that hide the original exception?

File: liquidctl/driver/smbus.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        return list(devs)
+
+    def __init__(self, smbus, description, vendor_id=None, product_id=None,
+                 address=None, **kwargs):
+        assert vendor_id and product_id and address is not None
+
+        self._smbus = smbus
+        self._description = description
+        self._vendor_id = vendor_id
+        self._product_id = product_id
```
Comment (MarshallAsch) [inline liquidctl/driver/smbus.py:220]: Should a note be added somewhere that these id's are actually the subsystem vendor and the subsystem product to remove some confusion when people are trying to make other drivers later?

File: docs/developer/style-guide.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+- use `key=value`, `` `expression` ``, `'string'`, or `value`, whichever is clearer
+
+### A quick refresher on `logging`
+
+Get a logger at the start of the module.
+
+```py
+import logging
+...
+_LOGGER = logging.getLogger(__name__)
```
Comment (MarshallAsch) [inline docs/developer/style-guide.md:100]: I think this recommended format of the logged name needs to be re-factored in a few of the other drivers as well, but that can be a separate PR.

File: liquidctl/driver/smbus.py (hunk 10 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/smbus.py:102]: I also only found out about this yesterday, despite it being documented in the introduction of [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html).

TL;DR: when using `from None` the original exception is kept in `__context__`, but is otherwise hidden; this avoids the display of double tracebacks.

File: liquidctl/driver/nvidia.py (hunk 9 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:132]: We could emit a warning here, but since it's unlikely to ever become a hard error (we're not going to find/expose new LED channels), I don't think it's necessary.

The reason why it wouldn't be a hard error is that I tend to prefer automatically correcting user input when that's possible/safe.  It's the same idea with when the user supplies too many colors.

File: liquidctl/driver/nvidia.py (hunk 8 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:1]: The basic issue is that I don't think it's sensible to have a bunch of single-driver modules, especially when there are significant similarities between the drivers.  _(Module-hell is also not idiomatic in Python)._

Eventually the we'll need to have one module for each "generation", so each module has one to ten drivers (each board partner has a few variants of each GPU, sometimes with different protocols).  But since we're going to be much more strict with adding these drivers, because of the increased associated risks, I'm betting it's going to take a long time until we have more than ten NVIDIA drivers in total.

For now I think we should keep all of them in a single `nvidia` module.  When that starts to get too big we'll split it into `pascal`, `turing`, `ampere`, etc., and at that point `nvidia` can simply re-export all drivers in a single namespace, which is more convenient for API users and also retains backwards compatibility.

File: liquidctl/driver/nvidia.py (hunk 7 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:45]: I had noticed a pattern and though maybe some additional modes could be uncovered.  But that didn't turn out to be the case, so I'll switch to hex for simplicity.

File: liquidctl/driver/nvidia.py (hunk 6 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/nvidia.py:53]: I guess it's a personal preference: the value is not needed outside of this function, so I decided to keep it local.

File: docs/nvidia-guide.md (hunk 4 above)
Comment (jonasmalacofilho) [reply docs/nvidia-guide.md:20]: It's intentional.  In this example of adding support for the 1070 FTW,

```
- [EVGA GTX 1080 FTW](#evga-gtx-1080-ftw)
```

is replaced with

```
- [EVGA GTX 1070 FTW](#evga-gtx-1070-1080-ftw)
- [EVGA GTX 1080 FTW](#evga-gtx-1070-1080-ftw)
```

The idea is to have both cards of the same "family" (protocol/driver) linking to the same "EVGA GTX 1070/1080 subsection".

File: liquidctl/driver/smbus.py (hunk 11 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/smbus.py:220]: Added.

File: docs/developer/style-guide.md (hunk 12 above)
Comment (jonasmalacofilho) [reply docs/developer/style-guide.md:100]: I think most are already doing that, but the logger instance is implicitly public (`LOGGER` instead of `_LOGGER`).

It's arguable whether we can change that now, but we can try...

More generally, the drivers are not consistent in coding-style, and that should be tackled sometime.

**Comment by jonasmalacofilho** [comment]
Thanks for the review!

For my part merging is currently blocked by:

- [ ] missing tests
- [ ] missing udev rules (will use yours, thanks!)
- [ ] missing `i2c-dev` loading shenanigans

File: liquidctl/driver/nvidia.py (hunk 9 above)
Comment (MarshallAsch) [reply liquidctl/driver/nvidia.py:132]: That makes sense.

File: liquidctl/driver/smbus.py (hunk 10 above)
Comment (MarshallAsch) [reply liquidctl/driver/smbus.py:102]: Would that be something that should be refactored elsewhere in other modules?

---
I will test that out a little to see how it shows expresses itself to learn that.

File: liquidctl/driver/nvidia.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
+"""liquidctl drivers for NVIDIA graphics cards.
+
+Copyright (C) 2020–2020  Jonas Malaco and contributors
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+from enum import Enum, unique
+import logging
+import re
```
Comment (MarshallAsch) [inline liquidctl/driver/nvidia.py:9]: unused import.

File: liquidctl/driver/smbus.py (hunk 10 above)
Comment (jonasmalacofilho) [reply liquidctl/driver/smbus.py:114]: Yes, and we should also switch to custom exceptions when applicable.
